Kuadrant/multi-cluster-traffic-controller
# Reconcile Istio Gateway

closes #45, closes #46 

*Steps to verify locally using Route53*

**IMPORTANT** These steps assume your local ./aws-credentials.env file has all values (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`) set and ./controller-config.env file has the right values for `ZONE_ROOT_DOMAIN` and `AWS_DNS_PUBLIC_ZONE_ID` that match dev.hcpapps.net.

**IMPORTANT** Change the hostname in the Gateway and HttpRoute to avoid reusing `myapp.dev.hcpapps.net` e.g. Use `<myname>.dev.hcpapps.net`

```
cp config/samples/gateway.yaml ./gateway.yaml
cp config/samples/echo_httproute.yaml ./echo_httproute.yaml
vi gateway.yaml
vi echo_httproute.yaml
```

**Terminal 1** Local setup & start the controller in the control plane cluster
```
MCTC_WORKLOAD_CLUSTERS_COUNT=1 make local-setup
(export $(cat ./controller-config.env | xargs) && export $(cat ./aws-credentials.env | xargs) && make build-controller install run-controller
```

**Terminal 2** Start the syncer in the workload cluster
```
export KUBECONFIG=./tmp/kubeconfigs/mctc-workload-1.kubeconfig
make kind-load-syncer
make deploy-syncer
```

**Terminal 3** Create the GatewayClass & Gateway
```
export KUBECONFIG=./tmp/kubeconfigs/mctc-control-plane.kubeconfig
kubectl apply -f config/samples/gatewayclass.yaml
kubectl create namespace mctc-tenant
kubectl apply -n mctc-tenant -f ./gateway.yaml
```

**Terminal 2** Deploy the sample application with a HttpRoute
```
kubectl apply -n mctc-downstream -f ./echo_httproute.yaml
```

**Terminal 3** Wait for the Gateway to be programmed, then curl the hostname specified in the HttpRoute. You should get a 200 response.

```
kubectl wait --for=condition=Programmed --timeout=60s gateway example-gateway -n mctc-tenant
curl -k -s -o /dev/null -w "%{http_code}"  https://<myname>.dev.hcpapps.net
```

TODO: 

- [x] Rebase after #60 (#35)- Refactor zone selection/validation 
- [x] Rebase after sync agent spec #89 (#43) - Implement sync annotation setting
- [x] Rebase after sync agent status #96 (#43) - Implement status retrieval
- [x] Working integration test
- [x] Decouple from ingress controller that the ingress agent uses (initally re-used some of that code for a steel thread)
- [x] Fetch Gateway IP address(es) from Gateway status in `GetDNSTargets`
- [x] Check Gateway status for attached routes before creating DNS record (internal discussion link about attachedRoutes field in GatewayStatus https://redhat-internal.slack.com/archives/C047YHJJG2D/p1676550730816259)

Out of scope/follow up
- #85
- #52 
- Rebase after cluster registration changes #44 - Gateway cluster selector (sync to all clusters initially)
